ENH: sparse: add count_nonzero method

perimosocordiae opened the pull request:
This is a followup of #3524 by @larsmans.

Tests cover all formats, including non-canonical cases. There are currently two problems I'd like some feedback on:

 - <strike>Should I copy the docstring from the `spmatrix` implementation to each of the subclass' implementations? Also, is the docstring appropriate?</strike>
 - <strike>Non-canonical matrices currently fall back to the `spmatrix` generic implementation, which performs densification. I did this because it seemed like a `count_nonzero` method shouldn't ever mutate state, and the CSR/CSC `eliminate_zeros` method is in-place. Perhaps this case could be handled by making a copy and eliminating zeros on that?</strike>

perimosocordiae commented:
Bump?

pv commented on scipy/sparse/dok.py:
```diff
@@ -115,6 +115,9 @@ def getnnz(self):
     nnz = property(fget=getnnz)
 
+    def count_nonzero(self):
+        return sum(x != 0 for x in self.values())
```
six.itervalues

pv commented on scipy/sparse/base.py:
```diff
@@ -140,6 +140,19 @@ def getmaxprint(self):
+        actual number of non-zero entries in data.
+        """
+        # Subclasses may implement more efficient versions of this method,
+        # but this acts as a generic fallback in case they do not.
```
Do we have subclasses that don't implement this?
If not, maybe prefer to raise NotImplementedError?

pv commented on scipy/sparse/data.py:
```diff
@@ -66,6 +66,11 @@ def conj(self):
 
+    def count_nonzero(self):
+        if self.has_canonical_format:
+            return np.count_nonzero(self.data)
```
Maybe should `sum_duplicates`?

pv commented on scipy/sparse/dia.py:
```diff
@@ -151,6 +151,15 @@ def __repr__(self):
+        mask = (row >= 0)
+        mask &= (row < num_rows)
+        mask &= (offset_inds < num_cols)
+        return np.count_nonzero(self.data[mask])
```
Looks like it would be best to add `count_nonzero` tests with non-square matrices (both tall and wide...)

perimosocordiae replied:
The default test matrix has shape (3,4), so I added a test using its transpose as well.

pv commented:
Now that we more or less decided what to do, maybe push this forward again. LGTM mainly. Is `count_nonzero()` a good name (maybe, following Numpy) --- what did Julia choose?

As per the previous discussions, we opted for canonicalizing when necessary.

The documentation probably should be updated a bit, to indicate `nnz` are *structural nonzeros*, `count_zeros()` are actual nonzeros. These should go to the respective docstrings (for nnz, there's a property docstring?). It's probably OK to just do `count_nonzeros.__doc__ = spmatrix.count_nonzeros.__doc__` to avoid repeating the same docstring.

perimosocordiae commented:
I think I've addressed all your comments, @pv. The most recent commit also:

- deduplicates docstrings for `getnnz()`, `count_nonzero()`, and the `nnz` property.
- adds the `axis` kwarg to all `getnnz()` methods, with appropriate errors where not implemented.
- uses a separate docstring for the `nnz` property, to avoid mentioning the `axis` kwarg.

pv commented:
thanks, merged
